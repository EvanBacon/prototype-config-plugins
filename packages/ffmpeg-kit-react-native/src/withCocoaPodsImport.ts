import { mergeContents } from "@expo/config-plugins/build/utils/generateCode";
import {
  ConfigPlugin,
  withDangerousMod,
  withPodfileProperties,
} from "expo/config-plugins";
import { promises } from "fs";
import path from "path";

export const withPodfilePropertiesPackage: ConfigPlugin<string> = (
  config,
  packageName,
) => {
  return withPodfileProperties(config, (config) => {
    // @ts-ignore: wrong type
    config.modResults["ffmpeg-kit-react-native.subspecs"] = [
      packageName,
    ].filter(Boolean);
    return config;
  });
};

/** Dangerously adds the custom import to the CocoaPods. */
export const withCocoaPodsImport: ConfigPlugin = (c) => {
  return withDangerousMod(c, [
    "ios",
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    async (config) => {
      const file = path.join(config.modRequest.platformProjectRoot, "Podfile");

      const contents = await promises.readFile(file, "utf8");

      await promises.writeFile(file, addCocoaPodsImport(contents), "utf-8");
      return config;
    },
  ]);
};

export function addCocoaPodsImport(src: string): string {
  return mergeContents({
    tag: `ffmpeg-kit-react-native-import`,
    src,
    newSrc: `  pod 'ffmpeg-kit-react-native', :subspecs => podfile_properties['ffmpeg-kit-react-native.subspecs'] || [], :podspec => File.join(File.dirname(\`node --print "require.resolve('ffmpeg-kit-react-native/package.json')"\`), "ffmpeg-kit-react-native.podspec")`,
    anchor: /use_native_modules/,
    // We can't go after the use_native_modules block because it might have parameters, causing it to be multi-line (see react-native template).
    offset: 0,
    comment: "#",
  }).contents;
}
