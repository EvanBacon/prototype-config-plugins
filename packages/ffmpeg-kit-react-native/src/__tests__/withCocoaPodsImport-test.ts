import { getFixture } from "../../../../fixtures/getFixtures";
import { addCocoaPodsImport } from "../withCocoaPodsImport";

describe(addCocoaPodsImport, () => {
  it(`adds import`, () => {
    expect(addCocoaPodsImport(getFixture("Podfile"))).toMatchSnapshot();
  });
});
