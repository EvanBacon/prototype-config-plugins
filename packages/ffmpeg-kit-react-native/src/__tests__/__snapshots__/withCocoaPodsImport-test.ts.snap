// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addCocoaPodsImport adds import 1`] = `
"
platform :ios, '12.1'

require 'json'
podfile_properties = JSON.parse(File.read('./Podfile.properties.json')) rescue {}

target 'yolo68' do
  use_expo_modules!
# @generated begin ffmpeg-kit-react-native-import - expo prebuild (DO NOT MODIFY) sync-a79a962553e9fe0961383adf881bd0dada86e29b
  pod 'ffmpeg-kit-react-native', :subspecs => podfile_properties['ffmpeg-kit-react-native.subspecs'] || [], :podspec => File.join(File.dirname(\`node --print "require.resolve('ffmpeg-kit-react-native/package.json')"\`), "ffmpeg-kit-react-native.podspec")
# @generated end ffmpeg-kit-react-native-import
  config = use_native_modules!

  use_react_native!(
    :path => config[:reactNativePath],
    :hermes_enabled => podfile_properties['expo.jsEngine'] == 'hermes'
  )

  post_install do |installer|
    react_native_post_install(installer)

    installer.pods_project.targets.each do |target|
      if (target.name&.eql?('FBReactNativeSpec'))
        target.build_phases.each do |build_phase|
          if (build_phase.respond_to?(:name) && build_phase.name.eql?('[CP-User] Generate Specs'))
            target.build_phases.move(build_phase, 0)
          end
        end
      end
    end
  end
end
"
`;
